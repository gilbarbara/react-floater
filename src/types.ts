import * as React from 'react';
import { Data, Modifiers, Placement } from 'popper.js';
import { PartialDeep, RequireExactlyOne, ValueOf } from 'type-fest';

export type Action = 'open' | 'close';
export type HandlerFunction<T = HTMLElement> = (event: React.SyntheticEvent<T>) => void;
export type PlacementOptions = Placement | 'center';
export type PlainObject<T = any> = Record<string, T>;
export type Statuses = ValueOf<Status>;
export type SelectorOrElement = string | null | HTMLElement;

export interface Status {
  INIT: 'init';
  IDLE: 'idle';
  OPENING: 'opening';
  OPEN: 'open';
  CLOSING: 'closing';
  ERROR: 'error';
}

export interface RenderProps {
  closeFn: HandlerFunction;
}

export interface BaseProps {
  /* Open the Floater automatically. */
  autoOpen?: boolean;
  /* It will be called when the Floater change state */
  callback?: (action: Action, props: Props) => void;
  /* An element to trigger the Floater. */
  children?: React.ReactNode;
  /**
   * A React component or function to as a custom UI for the Floater.
   * The prop closeFloater will be available in your component.
   */
  component: React.FunctionComponent<RenderProps> | React.ReactElement;
  /**
   * The Floater content. It can be anything that can be rendered.
   * This is the only required props, unless you pass a component.
   */
  content: React.ReactNode;
  /* Log some basic actions. */
  debug?: boolean;
  /* Animate the Floater on scroll/resize. */
  disableAnimation?: boolean;
  /* Disable changes in the Floater position on scroll/resize. */
  disableFlip?: boolean;
  /* Don't convert hover event to click on mobile. */
  disableHoverToClick?: boolean;
  /* The event that will trigger the Floater. It can be hover | click. * These won't work in controlled mode. */
  event?: 'click' | 'hover';
  /* The amount of time (in seconds) that the floater should wait after a mouseLeave event before hiding. Only valid for event type hover  */
  eventDelay?: number;
  /* It can be anything that can be rendered. */
  footer?: React.ReactNode;
  /* Get the popper.js instance */
  getPopper?: (popper: Data, origin: 'floater' | 'wrapper') => void;
  /* Don't show the arrow. Useful for centered or modal layout. */
  hideArrow?: boolean;
  /* In case that you need to identify the portal. */
  id?: string;
  /* The distance between the Floater and its target in pixels. */
  offset?: number;
  /* Controlled mode. */
  open?: boolean;
  /* Customize popper.js modifiers. */
  options?: Modifiers;
  /* The placement of the Floater. It will update the position if there's no space available. */
  placement?: PlacementOptions;
  /* It will show a â¨‰ button to close the Floater. */
  showCloseButton?: boolean;
  style?: React.CSSProperties;
  /* Customize the default UI. */
  styles?: PartialDeep<Styles>;
  /* The target for the Floater positioning. If it's not set, it will use the `children` as the target. */
  target?: SelectorOrElement;
  /* It can be anything that can be rendered. */
  title?: React.ReactNode;
  /* Position the wrapper relative to the target. */
  wrapperOptions?: {
    offset?: number;
    placement?: Placement;
    position?: boolean;
  };
}

export type Props = RequireExactlyOne<BaseProps, 'content' | 'component'>;

export interface State {
  currentPlacement: PlacementOptions;
  positionWrapper: boolean;
  status: Statuses;
  statusWrapper: Statuses;
}

export interface CustomComponent {
  children?: React.ReactNode;
  closeFn: HandlerFunction;
}

export interface Styles {
  arrow: React.CSSProperties & {
    length: number;
    spread: number;
  };
  close: React.CSSProperties;
  container: React.CSSProperties;
  content: React.CSSProperties;
  floater: React.CSSProperties;
  floaterOpening: React.CSSProperties;
  floaterWithAnimation: React.CSSProperties;
  floaterWithComponent: React.CSSProperties;
  floaterClosing: React.CSSProperties;
  floaterCentered: React.CSSProperties;
  footer: React.CSSProperties;
  options: {
    zIndex: number;
  };
  title: React.CSSProperties;
  wrapper: React.CSSProperties;
  wrapperPosition: React.CSSProperties;
}
