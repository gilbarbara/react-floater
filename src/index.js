import React from 'react';
import PropTypes from 'prop-types';
import Popper from 'popper.js';
import deepmerge from 'deepmerge';
import { canUseDOM, isFixed, isMobile, isNode, once } from './utils';

import Tooltip from './Tooltip';

import stylesDefault from './styles';

const STATUS = {
  IDLE: 'idle',
  READY: 'ready',
  OPENING: 'opening',
  OPEN: 'open',
  CLOSING: 'closing',
  ERROR: 'error',
};

export default class ReactTooltips extends React.Component {
  constructor(props) {
    super(props);

    if (process.env.NODE_ENV !== 'production') {
      if (props.wrapperOptions.position && !props.target) {
        console.warn('Missing props! You need to set a `target` to use `wrapperOptions.position`'); //eslint-disable-line no-console
      }
    }

    this.state = {
      currentPlacement: props.placement,
      positionWrapper: props.wrapperOptions.position && !!props.target,
      status: STATUS.IDLE,
      statusWrapper: STATUS.IDLE,
    };
  }

  static propTypes = {
    animate: PropTypes.bool,
    autoOpen: PropTypes.bool,
    callback: PropTypes.func,
    children: PropTypes.node,
    content: PropTypes.node.isRequired,
    event: PropTypes.oneOf(['hover', 'click']),
    eventDelay: PropTypes.number,
    flip: PropTypes.bool,
    footer: PropTypes.node,
    id: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    offset: PropTypes.number,
    open: PropTypes.bool,
    placement: PropTypes.oneOf([
      'top', 'top-start', 'top-end',
      'bottom', 'bottom-start', 'bottom-end',
      'left', 'left-start', 'left-end',
      'right', 'right-start', 'right-end',
      'auto', 'center',
    ]),
    showCloseButton: PropTypes.bool,
    styles: PropTypes.object,
    target: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.string,
    ]),
    title: PropTypes.node,
    wrapperOptions: PropTypes.shape({
      placement: PropTypes.oneOf([
        'top', 'top-start', 'top-end',
        'bottom', 'bottom-start', 'bottom-end',
        'left', 'left-start', 'left-end',
        'right', 'right-start', 'right-end',
        'auto',
      ]),
      position: PropTypes.bool,
      offset: PropTypes.number,
    }),
    classNames: PropTypes.shape({
      arrow: PropTypes.string,
      container: PropTypes.string,
      content: PropTypes.string,
      footer: PropTypes.string,
      title: PropTypes.string,
      wrapper: PropTypes.string,
    }),
  };

  static defaultProps = {
    animate: true,
    autoOpen: false,
    callback: () => {},
    event: 'click',
    eventDelay: 0.4,
    flip: true,
    offset: 15,
    placement: 'bottom',
    showCloseButton: false,
    styles: {},
    target: null,
    wrapperOptions: {
      position: false,
    },
    classNames: {
      wrapper: '__tooltip',
      container: '__tooltip__container',
      content: '__tooltip__content',
      title: '__tooltip__title',
      footer: '__tooltip__footer',
      arrow: '__tooltip__arrow',
    },
  };

  componentDidMount() {
    if (!canUseDOM) return;

    this.initPopper();
  }

  componentWillReceiveProps(nextProps) {
    if (!canUseDOM) return;

    const { open, target, wrapperOptions } = this.props;

    if (open !== nextProps.open) {
      this.toggle();
    }

    if (wrapperOptions.position !== nextProps.wrapperOptions.position || target !== nextProps.target) {
      this.changeWrapperPosition(nextProps);
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (!canUseDOM) return;

    const { status } = this.state;
    const { autoOpen, open } = this.props;

    if (prevState.status === STATUS.IDLE && status === STATUS.READY) {
      if (autoOpen || open) {
        this.toggle(STATUS.OPEN);
      }
    }

    if (
      this.tooltip
      && ((prevState.status !== STATUS.OPENING && status === STATUS.OPENING)
        || (prevState.status !== STATUS.CLOSING && status === STATUS.CLOSING))
    ) {
      once(this.tooltip, 'transitionend', this.handleTransitionEnd);
    }
  }

  componentWillUnmount() {
    if (!canUseDOM) return;

    if (this.popper) {
      this.popper.instance.destroy();
    }

    if (this.wrapperPopper) {
      this.wrapperPopper.instance.destroy();
    }
  }

  initPopper(target = this.target) {
    const { positionWrapper } = this.state;
    const { flip, offset, placement, wrapperOptions } = this.props;
    const flipBehavior = placement === 'top' || placement === 'bottom' ? 'flip' : [
      'right',
      'bottom-end',
      'top-end',
      'left',
      'top-start',
      'bottom-start',
    ];

    if (placement === 'center') {
      this.setState({ status: STATUS.READY });
    }
    else if (target && this.tooltip) {
      new Popper(target, this.tooltip, {
        placement,
        modifiers: {
          arrow: {
            element: this.arrow,
          },
          offset: {
            offset: `0, ${offset}px`,
          },
          preventOverflow: {
            padding: 10,
          },
          flip: {
            enabled: flip,
            behavior: flipBehavior,
            padding: 20,
          }
        },
        onCreate: (data) => {
          this.popper = data;

          this.setState({
            currentPlacement: data.placement,
            status: STATUS.READY,
          });

          if (placement !== data.placement) {
            setTimeout(() => {
              data.instance.update();
            }, 1);
          }
        },
        onUpdate: (data) => {
          this.popper = data;

          if (data.placement !== this.state.currentPlacement) {
            this.setState({ currentPlacement: data.placement });
          }
        }
      });
    }

    if (positionWrapper) {
      const wrapperOffset = typeof wrapperOptions.offset !== 'undefined' ? wrapperOptions.offset : 0;

      new Popper(this.target, this.wrapper, {
        placement: wrapperOptions.placement || placement,
        modifiers: {
          arrow: {
            enabled: false,
          },
          offset: {
            offset: `0, ${wrapperOffset}px`,
          },
          flip: {
            enabled: false,
          }
        },
        onCreate: (data) => {
          this.wrapperPopper = data;
          this.setState({ statusWrapper: STATUS.READY });

          if (placement !== data.placement) {
            setTimeout(() => {
              data.instance.update();
            }, 1);
          }
        }
      });
    }
  }

  changeWrapperPosition({ target, wrapperOptions }) {
    this.setState({
      positionWrapper: wrapperOptions.position && !!target
    });
  }

  toggle(forceState) {
    let status = this.state.status === STATUS.OPEN ? STATUS.CLOSING : STATUS.OPENING;

    if (typeof forceState !== 'undefined') {
      status = forceState;
    }

    this.setState({ status });
  }

  setRef = (ref) => {
    this.tooltip = ref;
  };

  handleTransitionEnd = () => {
    const { callback } = this.props;

    if (this.wrapperPopper) {
      this.wrapperPopper.instance.update();
    }

    this.setState({
      status: this.state.status === STATUS.OPENING ? STATUS.OPEN : STATUS.READY
    }, () => {
      callback(this.state.status === STATUS.OPEN ? 'open' : 'close', this.props);
    });
  };

  handleClick = () => {
    if (typeof this.props.open !== 'undefined') return;

    const { positionWrapper } = this.state;

    if (
      this.eventType === 'click'
      || (positionWrapper && this.eventType === 'hover')
    ) {
      this.toggle();
    }
  };

  handleMouseEnter = () => {
    if (typeof this.props.open !== 'undefined') return;

    if (this.eventType === 'hover') {
      clearTimeout(this.eventDelayTimeout);
      this.toggle();
    }
  };

  handleMouseLeave = () => {
    const { status, positionWrapper } = this.state;
    const { eventDelay, open } = this.props;

    if (typeof open !== 'undefined') return;

    if (
      this.eventType === 'hover'
      && [STATUS.OPENING, STATUS.OPEN].includes(status)
      && !positionWrapper
    ) {
      if (!eventDelay) {
        this.toggle();
      }
      else {
        this.eventDelayTimeout = setTimeout(() => {
          this.toggle();
        }, eventDelay * 1000);
      }
    }
  };

  get eventType() {
    const { event } = this.props;

    if (event === 'hover' && isMobile()) {
      return 'click';
    }

    return event;
  }

  get target() {
    const { target } = this.props;
    let element = this.wrapper;

    if (target) {
      if (isNode(target)) {
        element = target;
      }
      else {
        element = document.querySelector(target);
      }
    }

    return element;
  }

  get styles() {
    const { status, positionWrapper, statusWrapper } = this.state;
    const { styles } = this.props;

    const combinedStyles = deepmerge(stylesDefault, styles);

    if (positionWrapper) {
      let wrapperStyle = {};

      if (![STATUS.READY].includes(status) || ![STATUS.READY].includes(statusWrapper)) {
        wrapperStyle = combinedStyles.wrapperPosition;
      }
      else {
        wrapperStyle = this.wrapperPopper.styles;
      }

      combinedStyles.wrapper = {
        ...combinedStyles.wrapper,
        ...wrapperStyle
      };
    }

    return combinedStyles;
  }

  get tooltipStyle() {
    const { currentPlacement, status } = this.state;
    const { animate } = this.props;
    const {
      arrow: { length },
      tooltip,
      tooltipCentered,
      tooltipClosing,
      tooltipOpening,
      tooltipWithAnimation,
    } = this.styles;
    let styles = {};

    if (currentPlacement.startsWith('top')) {
      styles.padding = `0 10px ${length}px`;
    }
    else if (currentPlacement.startsWith('bottom')) {
      styles.padding = `${length}px 10px 0`;
    }
    else if (currentPlacement.startsWith('left')) {
      styles.padding = `0 ${length}px 0 0`;
    }
    else if (currentPlacement.startsWith('right')) {
      styles.padding = `0 0 0 ${length}px`;
    }

    if ([STATUS.OPENING, STATUS.OPEN].includes(status)) {
      styles = { ...styles, ...tooltipOpening };
    }

    if (status === STATUS.CLOSING) {
      styles = { ...styles, ...tooltipClosing };
    }

    if (status === STATUS.OPEN && animate && !isFixed(this.target)) {
      styles = { ...styles, ...tooltipWithAnimation };
    }

    if (currentPlacement === 'center') {
      styles = { ...styles, ...tooltipCentered };
    }

    return {
      ...tooltip,
      ...styles,
    };
  }

  get arrowStyle() {
    const { currentPlacement } = this.state;
    const { length } = this.styles.arrow;
    const styles = {
      position: 'absolute',
    };

    if (currentPlacement.startsWith('top')) {
      styles.bottom = length;
      styles.left = 0;
      styles.right = 0;
    }
    else if (currentPlacement.startsWith('bottom')) {
      styles.top = 0;
      styles.left = 0;
      styles.right = 0;
    }
    else if (currentPlacement.startsWith('left')) {
      styles.right = 0;
      styles.top = 0;
      styles.bottom = 0;
    }
    else if (currentPlacement.startsWith('right')) {
      styles.left = 0;
      styles.top = 0;
    }

    return styles;
  }

  renderArrow() {
    const { currentPlacement } = this.state;
    const { arrow: { color, display, length, position, spread } } = this.styles;
    const styles = { display, position };

    let points;
    let x = spread;
    let y = length;

    if (currentPlacement.startsWith('top')) {
      points = `0,0 ${x / 2},${y} ${x},0`;
      styles.bottom = -y;
    }
    else if (currentPlacement.startsWith('bottom')) {
      points = `${x},${y} ${x / 2},0 0,${y}`;
      styles.top = 0;
    }
    else if (currentPlacement.startsWith('left')) {
      y = spread;
      x = length;
      points = `0,0 ${x},${y / 2} 0,${y}`;
      styles.right = 0;
    }
    else if (currentPlacement.startsWith('right')) {
      y = spread;
      x = length;
      points = `${x},${y} ${x},0 0,${y / 2}`;
      styles.left = 0;
    }

    return (
      <span ref={c => (this.arrow = c)} style={styles}>
        <svg
          width={x}
          height={y}
          version="1.1"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon points={points} fill={color} />
        </svg>
      </span>
    );
  }

  renderTooltip() {
    const { currentPlacement, positionWrapper } = this.state;
    const { content, footer, open, showCloseButton, title, classNames } = this.props;

    const { styles } = this;

    const output = {
      content: React.isValidElement(content)
        ? content
        : <div className={classNames.content} style={styles.content}>{content}</div>
    };

    if (title) {
      output.title = React.isValidElement(title)
        ? title
        : <div className={classNames.title} style={styles.title}>{title}</div>;
    }

    if (footer) {
      output.footer = React.isValidElement(footer)
        ? footer
        : <div className={classNames.footer} style={styles.footer}>{footer}</div>;
    }

    if (
      (showCloseButton || positionWrapper)
      && typeof open === 'undefined'
    ) {
      output.close = (<button style={styles.close} onClick={this.handleClick}>×︎️</button>);
    }

    if (currentPlacement !== 'center') {
      output.arrow = (
        <div
          className={classNames.arrow}
          style={this.arrowStyle}
        >
          {this.renderArrow()}
        </div>
      );
    }

    return (
      <div
        ref={c => (this.tooltip = c)}
        className={classNames.wrapper}
        style={this.tooltipStyle}
      >
        <div className={classNames.container} style={styles.container}>
          {output.close}
          {output.title}
          {output.content}
          {output.footer}
        </div>
        {output.arrow}
      </div>
    );
  }

  renderWrapper() {
    const { children } = this.props;
    const { wrapper } = this.styles;

    return (
      <span
        key="wrapper"
        ref={c => (this.wrapper = c)}
        style={wrapper}
        onClick={this.handleClick}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
      >
        {children}
      </span>
    );
  }

  render() {
    return (
      <Tooltip
        {...this.props}
        placement={this.state.currentPlacement}
        setRef={this.setRef}
        status={this.state.status}
        tooltip={this.renderTooltip()}
        wrapper={this.renderWrapper()}
      />);
  }
}
