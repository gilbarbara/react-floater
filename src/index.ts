import ReactFloater from './components';

export default ReactFloater;
