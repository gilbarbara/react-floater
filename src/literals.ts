export const POSITIONING_PROPS = ['position', 'top', 'right', 'bottom', 'left'] as const;

export const STATUS = {
  INIT: 'init',
  IDLE: 'idle',
  RENDER: 'render',
  OPENING: 'opening',
  OPEN: 'open',
  CLOSING: 'closing',
  ERROR: 'error',
} as const;
