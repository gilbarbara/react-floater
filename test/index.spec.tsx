import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';

import ReactFloater from '../src';
import { Props, State } from '../src/types';

import Styled from './__fixtures__/Styled';

jest.useFakeTimers();

const mockCallback = jest.fn();
const mockGetPopper = jest.fn(() => ({ instance: {} }));

const props: Props = {
  content: 'Hello! This is my content!',
  getPopper: mockGetPopper,
};

function setup(
  ownProps = props,
  children: string | React.ReactNode = 'Places',
): ReactWrapper<Props, State> {
  return mount(<ReactFloater {...ownProps}>{children}</ReactFloater>);
}

describe('ReactFloater', () => {
  let portal: ReactWrapper;
  let floater: ReactWrapper<Props, State>;

  const updateTooltip = (event: string | false = 'click') => {
    if (event) {
      floater.find('ReactFloaterWrapper').childAt(0).simulate(event);
      const delay = floater.prop('eventDelay') || 1;

      if (['click', 'mouseEnter'].includes(event)) {
        // @ts-ignore
        floater.instance().handleTransitionEnd(); // mock transitionend
      } else {
        setTimeout(() => {
          // @ts-ignore
          floater.instance().handleTransitionEnd();
        }, delay * 1100);
      }
    }

    floater.update();
    portal = floater.find('ReactFloaterPortal');
  };

  describe('basic usage', () => {
    beforeAll(() => {
      floater = setup();
      portal = floater.find('ReactFloaterPortal');
    });

    it('should render properly', () => {
      expect(floater.find('ReactFloater')).toExist();
      expect(floater.find('ReactFloaterPortal').length).toBe(1);
      expect(floater.find('ReactFloaterWrapper span').at(0)).toHaveText('Places');
    });

    it('should have created a Portal', () => {
      expect(portal.find('.__floater')).toExist();
      expect(portal.find('.__floater__container')).toExist();
      expect(portal.find('.__floater__arrow')).toExist();
    });

    it('should have called getPopper', () => {
      // @ts-ignore
      const popper = mockGetPopper.mock.calls[0][0];

      // @ts-ignore
      expect(popper.instance.constructor.name).toBe('Popper');
    });

    it('should have rendered the Floater initially hidden', () => {
      const floaterEl = portal.find('.__floater');

      expect(floater.state('status')).toBe('idle');
      expect(floater.find('.__floater__container')).toHaveText('Hello! This is my content!');
      expect(floaterEl).toHaveStyle('opacity', 0);
      expect(floaterEl).toHaveStyle('visibility', 'hidden');
    });

    it('should be able to show the floater', () => {
      updateTooltip();
      const floaterEl = portal.find('.__floater');

      expect(floater.state('status')).toBe('open');
      expect(floaterEl).toHaveStyle('opacity', 1);
      expect(floaterEl).toHaveStyle('visibility', 'visible');
    });

    it('should be able to hide the floater', () => {
      updateTooltip();
      const floaterEl = portal.find('.__floater');

      expect(floater.state('status')).toBe('idle');
      expect(floaterEl).toHaveStyle('opacity', 0);
      expect(floaterEl).toHaveStyle('visibility', 'hidden');
    });

    it('should unmount properly', () => {
      floater.unmount();
      expect(floater.find('ReactFloater')).not.toExist();

      expect(floater.find('ReactFloaterPortal')).not.toExist();
    });
  });

  describe('with multiple children', () => {
    beforeAll(() => {
      floater = setup(props, [<div key={0}>Hello</div>, <div key={1}>World</div>]);
      portal = floater.find('ReactFloaterPortal');
    });

    it('should render properly', () => {
      const content = floater.find('ReactFloaterWrapper').childAt(0).find('div');

      expect(floater.find('ReactFloater')).toExist();
      expect(floater.find('ReactFloaterPortal')).toExist();
      expect(content.at(0)).toHaveText('Hello');
      expect(content.at(1)).toHaveText('World');
    });
  });

  describe('with `autoOpen`', () => {
    beforeAll(() => {
      floater = setup({
        ...props,
        autoOpen: true,
      });
    });

    it('should have rendered the Floater initially open', () => {
      updateTooltip(false);
      const floaterEl = portal.find('.__floater');

      expect(floater.state('status')).toBe('open');
      expect(floaterEl).toHaveStyle('opacity', 1);
      expect(floaterEl).toHaveStyle('visibility', 'visible');
    });

    it('should be able to hide the floater', () => {
      updateTooltip();
      const floaterEl = portal.find('.__floater');

      expect(floater.state('status')).toBe('idle');
      expect(floaterEl).toHaveStyle('opacity', 0);
      expect(floaterEl).toHaveStyle('visibility', 'hidden');
    });

    it('should be able to show the floater again', () => {
      updateTooltip();
      const floaterEl = portal.find('.__floater');

      expect(floater.state('status')).toBe('open');
      expect(floaterEl).toHaveStyle('opacity', 1);
      expect(floaterEl).toHaveStyle('visibility', 'visible');
    });
  });

  describe('with `callback`', () => {
    beforeAll(() => {
      floater = setup({
        ...props,
        callback: mockCallback,
      });
    });

    it('should call the callback function on open', () => {
      updateTooltip();

      expect(floater.state('status')).toBe('open');
      expect(mockCallback).toHaveBeenCalledWith('open', {
        autoOpen: false,
        callback: mockCallback,
        children: 'Places',
        content: 'Hello! This is my content!',
        debug: false,
        disableAnimation: false,
        disableFlip: false,
        disableHoverToClick: false,
        event: 'click',
        eventDelay: 0.4,
        getPopper: mockGetPopper,
        hideArrow: false,
        offset: 15,
        placement: 'bottom',
        showCloseButton: false,
        styles: {},
        target: null,
        wrapperOptions: { position: false },
      });
    });

    it('should call the callback function on close', () => {
      updateTooltip();
      expect(floater.state('status')).toBe('idle');

      expect(mockCallback).toHaveBeenCalledWith('close', {
        autoOpen: false,
        callback: mockCallback,
        children: 'Places',
        content: 'Hello! This is my content!',
        debug: false,
        disableAnimation: false,
        disableFlip: false,
        disableHoverToClick: false,
        event: 'click',
        eventDelay: 0.4,
        getPopper: mockGetPopper,
        hideArrow: false,
        offset: 15,
        placement: 'bottom',
        showCloseButton: false,
        styles: {},
        target: null,
        wrapperOptions: { position: false },
      });
    });
  });

  describe('with `event` hover', () => {
    beforeAll(() => {
      floater = setup({
        ...props,
        event: 'hover',
      });
    });

    it('should be able to show the floater', () => {
      updateTooltip('mouseEnter');

      expect(floater.state('status')).toBe('open');
    });

    it('should still be open while the cursor is over it', () => {
      updateTooltip(false);

      expect(floater.state('status')).toBe('open');
    });

    it('should have close itself after `eventDelay`', () => {
      updateTooltip('mouseLeave');

      // @ts-ignore
      jest.advanceTimersByTime(floater.prop('eventDelay') * 1000); // trigger the close animation
      expect(floater.state('status')).toBe('closing');

      jest.advanceTimersByTime(400); // trigger the fake transitionend event
      expect(floater.state('status')).toBe('idle');
    });
  });

  describe('with `event` hover and `eventDelay` set to 0', () => {
    beforeAll(() => {
      floater = setup({
        ...props,
        event: 'hover',
        eventDelay: 0,
      });
    });

    it('should be able to show the floater', () => {
      updateTooltip('mouseEnter');

      expect(floater.state('status')).toBe('open');
    });

    it('should have close itself immediately', () => {
      updateTooltip('mouseLeave');

      expect(floater.state('status')).toBe('idle');

      jest.advanceTimersByTime(0); // trigger the fake transitionend event

      expect(floater.state('status')).toBe('idle');
    });
  });

  describe('with `title`', () => {
    beforeAll(() => {
      const Title = () => <h3>My Title</h3>;
      floater = setup({
        ...props,
        title: <Title />,
      });
    });

    it('should have rendered the title', () => {
      expect(floater.find('Title')).toExist();

      floater.setProps({
        title: <div className="__title">Other Title</div>,
      });

      expect(floater.find('Title')).not.toExist();
      expect(floater.find('.__title')).toExist();
    });
  });

  describe('with `footer`', () => {
    beforeAll(() => {
      const Footer = () => (
        <footer>
          <button type="button">NEXT</button>
        </footer>
      );
      floater = setup({
        ...props,
        footer: <Footer />,
      });
    });

    it('should have rendered the footer', () => {
      expect(floater.find('Footer')).toExist();

      floater.setProps({
        footer: <div className="__footer">Hello</div>,
      });

      expect(floater.find('Footer')).not.toExist();
      expect(floater.find('.__footer')).toExist();
    });
  });

  describe('with `id`', () => {
    beforeAll(() => {
      floater = setup({
        ...props,
        id: 'hello-world',
      });
    });

    it('should have added the id to the portal floater', () => {
      const namedPortal = document.getElementById('hello-world');

      expect(namedPortal).toBeInstanceOf(HTMLDivElement);
      expect(namedPortal?.querySelector('.__floater')).toBeInstanceOf(HTMLDivElement);
    });
  });

  describe('with `open`', () => {
    beforeAll(() => {
      floater = setup({
        ...props,
        open: false,
      });
    });

    it('should not be able to show the floater with click', () => {
      updateTooltip('click');

      expect(floater.state('status')).toBe('idle');
    });

    it('should not be able to show the floater with hover', () => {
      floater.setProps({ event: 'hover' });
      updateTooltip('mouseEnter');

      expect(floater.state('status')).toBe('idle');
    });

    it('should show the floater when `open` is true', () => {
      floater.setProps({ open: true });
      expect(floater.state('status')).toBe('opening');

      // @ts-ignore
      floater.instance().handleTransitionEnd();
      expect(floater.state('status')).toBe('open');
    });

    it('should close the floater when `open` is false', () => {
      floater.setProps({ open: false });
      expect(floater.state('status')).toBe('closing');

      // @ts-ignore
      floater.instance().handleTransitionEnd();
      expect(floater.state('status')).toBe('idle');
    });
  });

  describe('with `placement` top', () => {
    beforeAll(() => {
      floater = setup({
        ...props,
        placement: 'top',
      });
    });

    it('should use `placement` top', () => {
      // @ts-ignore
      expect(floater.instance().popper.originalPlacement).toBe('top');
      expect(floater.state('currentPlacement')).toBe('top');
    });
  });

  describe('with `placement` center', () => {
    beforeAll(() => {
      floater = setup({
        ...props,
        placement: 'center',
      });
    });

    it('should use `placement` center', () => {
      // @ts-ignore
      expect(floater.instance().popper).not.toBeDefined();
      expect(floater.state('currentPlacement')).toBe('center');
    });

    it('should have skipped the arrow', () => {
      updateTooltip();

      const floaterEl = portal.find('.__floater');

      expect(floater.state('status')).toBe('open');
      expect(floaterEl.find('.__floater__arrow')).not.toExist();
    });
  });

  describe('with `component` as function', () => {
    beforeAll(() => {
      floater = setup({
        component: Styled,
      });
    });

    it('should show the floater with click', () => {
      updateTooltip('click');

      expect(floater.state('status')).toBe('open');
    });

    it('should have rendered the component', () => {
      expect(floater.find('.__floater__body').html()).toMatchSnapshot();
    });

    it('should be able to close the floater with `closeFn` prop', () => {
      floater.find('.__floater__body button').simulate('click');

      expect(floater.state('status')).toBe('closing');
    });
  });

  describe('with `component` as element', () => {
    beforeAll(() => {
      floater = setup({
        component: <Styled />,
        styles: {
          options: {
            zIndex: 1000,
          },
        },
      });
    });

    it('should show the floater with click', () => {
      updateTooltip('click');

      expect(floater.state('status')).toBe('open');
    });

    it('should have rendered the component', () => {
      expect(floater.find('.__floater__body').html()).toMatchSnapshot();
    });

    it('should be able to close the floater with `closeFn` prop', () => {
      floater.find('.__floater__body button').simulate('click');

      expect(floater.state('status')).toBe('closing');
    });
  });

  describe('with `showCloseButton`', () => {
    beforeAll(() => {
      floater = setup({
        ...props,
        showCloseButton: true,
      });
    });

    it('should show the floater with click', () => {
      updateTooltip('click');

      expect(floater.state('status')).toBe('open');
    });

    it('should have a close button', () => {
      expect(floater.find('FloaterCloseBtn')).toExist();
    });

    it('should be able to close the floater clicking the close button', () => {
      floater.find('FloaterCloseBtn').simulate('click');

      expect(floater.state('status')).toBe('closing');
    });
  });
});
