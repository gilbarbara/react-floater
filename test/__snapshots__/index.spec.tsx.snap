// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactFloater with \`component\` as element should have rendered the component 1`] = `
<div class="__floater__body">
  <div class="sc-bdfBwQ gbZmDx">
    <div>
      Styled
    </div>
    <button type="button">
      close
    </button>
  </div>
  <div class="__floater__arrow"
       style="pointer-events: none; position: absolute; width: 100%; left: 0px; right: 0px; top: 0px; height: 16px;"
  >
    <span style="display: inline-flex; position: absolute; margin-left: 8px; margin-right: 8px;">
      <svg width="32"
           height="16"
           version="1.1"
           xmlns="http://www.w3.org/2000/svg"
      >
        <polygon points="32,16 16,0 0,16"
                 fill="#fff"
        >
        </polygon>
      </svg>
    </span>
  </div>
</div>
`;

exports[`ReactFloater with \`component\` as function should have rendered the component 1`] = `
<div class="__floater__body">
  <div class="sc-bdfBwQ gbZmDx">
    <div>
      Styled
    </div>
    <button type="button">
      close
    </button>
  </div>
  <div class="__floater__arrow"
       style="pointer-events: none; position: absolute; width: 100%; left: 0px; right: 0px; top: 0px; height: 16px;"
  >
    <span style="display: inline-flex; position: absolute; margin-left: 8px; margin-right: 8px;">
      <svg width="32"
           height="16"
           version="1.1"
           xmlns="http://www.w3.org/2000/svg"
      >
        <polygon points="32,16 16,0 0,16"
                 fill="#fff"
        >
        </polygon>
      </svg>
    </span>
  </div>
</div>
`;
